File: packages/lhc/src/shared-tech/work-queue/claim-fence.ts
// Claim-attempt fencing and clean-exit handback.
//
// An expired claim is requeued in place (same work_item_id), so the id and
// status='claimed' no longer name one holder: a slow old holder and its retry
// would both match. Each claim bumps `claimAttempt` in the row's payload and
// the holder carries that number; every consume (success, terminal failure,
// deferral, delete, requeue) matches it too. Rows claimed before the counter
// existed carry none and their holders none, so NULL matches NULL.
//
// A process that exits cleanly hands back the claims it still holds (queued
// again, and any claimExpired mark cleared), so a normal one-shot exit
// mid-derivation never counts as an expiry; only a crash or kill expires.
import { existsSync } from "node:fs";
import { DatabaseSync } from "node:sqlite";
import { databasePathFor } from "../storage.js";

export interface ClaimAttempt {
  workItemId: string;
  claimAttempt?: number | undefined;
}

/** WHERE clause naming one claim attempt; bind with claimParams(). */
export const OWNED_CLAIM_SQL = `work_item_id = ? AND status = 'claimed' AND json_extract(payload, '$.claimAttempt') IS ?`;

export function claimParams(claim: ClaimAttempt): [string, number | null] {
  return [claim.workItemId, claim.claimAttempt ?? null];
}

/** The claim UPDATE's SET fragment: take the claim and start a new attempt. */
export const TAKE_CLAIM_SET_SQL = `status = 'claimed', claimed_at = ?, claim_expires_at = ?,
  payload = json_set(payload, '$.claimAttempt', COALESCE(json_extract(payload, '$.claimAttempt'), 0) + 1)`;

const held = new Map<string, { path: string; workItemId: string; claimAttempt: number | null }>();
let exitHookInstalled = false;

function heldKey(path: string, claim: ClaimAttempt): string {
  return `${path}\u0000${claim.workItemId}\u0000${claim.claimAttempt ?? ""}`;
}

/** Record a claim this process now holds (no-op for a handle without a known path). */
export function noteClaimHeld(db: DatabaseSync, claim: ClaimAttempt): void {
  const path = databasePathFor(db);
  if (path === undefined) return;
  held.set(heldKey(path, claim), { path, workItemId: claim.workItemId, claimAttempt: claim.claimAttempt ?? null });
  if (!exitHookInstalled) {
    exitHookInstalled = true;
    process.once("exit", releaseHeldClaims);
  }
}

/** The claim is consumed or lost; this process no longer holds it. */
export function noteClaimDone(db: DatabaseSync, claim: ClaimAttempt): void {
  const path = databasePathFor(db);
  if (path !== undefined) held.delete(heldKey(path, claim));
}

/**
 * Hand every still-held claim back to the queue. Runs from the process "exit"
 * event (synchronous, like node:sqlite). Fenced to the attempt, so a claim that
 * expired and was taken by another process is left alone. Best effort: a
 * failure leaves the claim to expire as before.
 */
export function releaseHeldClaims(): number {
  const byPath = new Map<string, Array<{ workItemId: string; claimAttempt: number | null }>>();
  for (const claim of held.values()) {
    const list = byPath.get(claim.path) ?? [];
    list.push(claim);
    byPath.set(claim.path, list);
  }
  held.clear();
  let released = 0;
  for (const [path, claims] of byPath) {
    // A removed thread file stays removed: opening it would recreate it empty.
    if (!existsSync(path)) continue;
    let db: DatabaseSync | undefined;
    try {
      db = new DatabaseSync(path);
      db.exec("PRAGMA busy_timeout = 2000;");
      const release = db.prepare(
        `UPDATE work_item SET status = 'queued', claimed_at = NULL, claim_expires_at = NULL,
           payload = json_remove(payload, '$.claimExpired')
         WHERE ${OWNED_CLAIM_SQL}`,
      );
      for (const claim of claims) {
        released += Number(release.run(claim.workItemId, claim.claimAttempt).changes);
      }
    } catch {
      // best effort: the claim expires as it did before
    } finally {
      try {
        db?.close();
      } catch {
        // best effort
      }
    }
  }
  return released;
}

File: packages/lhc-rs/src/shared_tech/work_queue/claim_fence.rs
//! Claim-attempt fencing and clean-exit handback.
//!
//! Ported from packages/lhc/src/shared-tech/work-queue/claim-fence.ts.
//! Node `process.once("exit")` is not used. Hosts call
//! [`release_held_claims_for`] when one thread closes or unloads, and
//! [`release_held_claims`] once at coordinated process shutdown.
//!
//! Rust-only scoped call: Codex serves many thread databases in one process,
//! so a process-wide take on thread close would requeue other live threads'
//! in-flight claims. TS keeps exit-only `releaseHeldClaims` (cc-lhc is one
//! thread per process). Fencing and the two-expiry policy are unchanged.
//!
//! Hand-back sqlite uses crate-private fallible storage (`open_database_for_handback`,
//! `exec_fallible` / `prepare_fallible` / `run_fallible` / `close_fallible`),
//! not the panicking adapter and not `catch_unwind`.

use std::collections::HashMap;
use std::sync::Mutex;

use super::super::storage::{Db, SqlParam, open_database_for_handback};

/// WHERE clause naming one claim attempt; bind with [`claim_params`].
pub const OWNED_CLAIM_SQL: &str =
    "work_item_id = ? AND status = 'claimed' AND json_extract(payload, '$.claimAttempt') IS ?";

/// The claim UPDATE's SET fragment: take the claim and start a new attempt.
pub const TAKE_CLAIM_SET_SQL: &str = "status = 'claimed', claimed_at = ?, claim_expires_at = ?,
  payload = json_set(payload, '$.claimAttempt', COALESCE(json_extract(payload, '$.claimAttempt'), 0) + 1)";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ClaimAttempt {
    pub work_item_id: String,
    pub claim_attempt: Option<i64>,
}

pub fn claim_params(claim: &ClaimAttempt) -> [SqlParam; 2] {
    [
        SqlParam::from(claim.work_item_id.as_str()),
        match claim.claim_attempt {
            Some(n) => SqlParam::I64(n),
            None => SqlParam::Null,
        },
    ]
}

struct HeldClaim {
    path: String,
    work_item_id: String,
    claim_attempt: Option<i64>,
}

fn held_key(path: &str, claim: &ClaimAttempt) -> String {
    format!(
        "{}\u{0000}{}\u{0000}{}",
        path,
        claim.work_item_id,
        claim
            .claim_attempt
            .map(|n| n.to_string())
            .unwrap_or_default()
    )
}

static HELD: Mutex<Option<HashMap<String, HeldClaim>>> = Mutex::new(None);

fn held_map() -> std::sync::MutexGuard<'static, Option<HashMap<String, HeldClaim>>> {
    HELD.lock().unwrap_or_else(|poisoned| poisoned.into_inner())
}

/// Record a claim this process now holds (no-op for a handle without a known path).
pub fn note_claim_held(db: &Db, claim: &ClaimAttempt) {
    let path = db.path();
    if path.is_empty() {
        return;
    }
    let key = held_key(path, claim);
    let mut guard = held_map();
    let map = guard.get_or_insert_with(HashMap::new);
    map.insert(
        key,
        HeldClaim {
            path: path.to_string(),
            work_item_id: claim.work_item_id.clone(),
            claim_attempt: claim.claim_attempt,
        },
    );
}

/// The claim is consumed or lost; this process no longer holds it.
pub fn note_claim_done(db: &Db, claim: &ClaimAttempt) {
    let path = db.path();
    if path.is_empty() {
        return;
    }
    let key = held_key(path, claim);
    if let Some(map) = held_map().as_mut() {
        map.remove(&key);
    }
}

fn warn_handback(path: &str, detail: &str) {
    eprintln!("lhc: release_held_claims failed for {path}: {detail} (claim left to expire)");
}

fn release_taken_claims(path: &str, list: Vec<(String, Option<i64>)>) -> i64 {
    if list.is_empty() || !std::path::Path::new(path).exists() {
        return 0;
    }
    let db = match open_database_for_handback(path) {
        Ok(db) => db,
        Err(detail) => {
            warn_handback(path, &detail);
            return 0;
        }
    };
    let released = match release_claims_on_db(&db, list) {
        Ok(n) => n,
        Err(detail) => {
            warn_handback(path, &detail);
            0
        }
    };
    if let Err(detail) = db.close_fallible() {
        warn_handback(path, &detail);
    }
    released
}

fn release_claims_on_db(db: &Db, list: Vec<(String, Option<i64>)>) -> Result<i64, String> {
    db.exec_fallible("PRAGMA busy_timeout = 2000;")?;
    let sql = format!(
        "UPDATE work_item SET status = 'queued', claimed_at = NULL, claim_expires_at = NULL,
           payload = json_remove(payload, '$.claimExpired')
         WHERE {OWNED_CLAIM_SQL}"
    );
    let stmt = db.prepare_fallible(&sql)?;
    let mut n = 0i64;
    for (work_item_id, claim_attempt) in list {
        n += stmt
            .run_fallible(&claim_params(&ClaimAttempt {
                work_item_id,
                claim_attempt,
            }))?
            .changes;
    }
    Ok(n)
}

/// Hand claims held on one thread database back to the queue. Fenced to the
/// attempt. Hosts call this when that thread closes or unloads; other
/// databases' in-flight claims stay held. Best effort: a failure is logged,
/// the connection is closed, and the unsuccessful claim is left to expire.
/// Never panics: sqlite failures are logged and skipped.
pub fn release_held_claims_for(path: &str) -> i64 {
    if path.is_empty() {
        return 0;
    }
    let list = {
        let mut guard = held_map();
        let Some(map) = guard.as_mut() else {
            return 0;
        };
        let mut list = Vec::new();
        map.retain(|_, claim| {
            if claim.path == path {
                list.push((claim.work_item_id.clone(), claim.claim_attempt));
                false
            } else {
                true
            }
        });
        list
    };
    release_taken_claims(path, list)
}

/// Hand every still-held claim back to the queue. Fenced to the attempt.
/// Process-wide: hosts invoke this once at coordinated process shutdown,
/// after stopping claim admission and settling or cancelling workers, not
/// on a single thread's close. Best effort per database: a failure is
/// logged, the connection is closed, remaining databases continue, and the
/// unsuccessful claim is left to expire as before. Never panics: sqlite
/// failures are logged and remaining databases continue.
pub fn release_held_claims() -> i64 {
    let claims = {
        let mut guard = held_map();
        guard.take().unwrap_or_default()
    };
    let mut by_path: HashMap<String, Vec<(String, Option<i64>)>> = HashMap::new();
    for claim in claims.into_values() {
        by_path
            .entry(claim.path)
            .or_default()
            .push((claim.work_item_id, claim.claim_attempt));
    }
    let mut released = 0i64;
    for (path, list) in by_path {
        released += release_taken_claims(&path, list);
    }
    released
}
